memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+========================
+
+Process initialization using dynamic table entries
+--------------------------------------------------
+
+During process intialization, the dyanmic loader should use the MemtagABI
+dynamic table entries to enable the correct MTE features at runtime.
+
+If ``DT_AARCH64_MEMTAG_MODE`` is present, the dynamic loader should (in a
+platform-specific specific way) enable MTE for the process. On Linux, this is
+done through ``prctl(PR_SET_TAGGED_ADDR_CTRL, PR_TAGGED_ADDR_ENABLE |
+PR_MTE_TCF_[A]SYNC ...)``. When ``DT_AARCH64_MEMTAG_MODE`` is ``0``, then
+synchronous MTE should be enabled. When ``DT_AARCH64_MEMTAG_MODE`` is ``1``,
+asynchronous MTE should be enabled.
+
+If ``DT_AARCH64_MEMTAG_HEAP`` is present, the dynamic loader should inform the
+heap allocator (in a platform-specific way) that memory tags should be used for
+all heap allocations in this process.
+
+If ``DT_AARCH64_MEMTAG_STACK`` is present, the dynamic loader should enable
+tagging for the main stack and thread stacks. Re-mapping currently alive stack
+frames as taggable is dangerous, and so dynamic loader developers may find it
+easiest to implement this functionality by:
+
+1. Ensuring the implementation completes before any threads have spawned, and
+
+2. Creating a new MTE-protected main stack and calling a function that
+   transitions onto the new stack, preserving all existing frames on the
+   untagged stack.
+
+Global tagging
+--------------
+
+Because memory tags can only be applied to tag-capable memory, the dynamic
+loader should ensure that any segments containing global variables are mapped as
+tag-capable. On linux, this is done by ensuring that ensuring that the mappings
+are created as anonymous mappings (``MAP_ANONYMOUS``) with the ``PROT_MTE``
+flag. Note that this requires special handling for when the linker is invoked
+implicitly (via. the value in ``PT_INTERP`` rather than as an argument to the
+loader binary), as the kernel will load the process into memory, rather than the
+loader. In these cases, the loader is required to unmap the already-loaded
+segments, and remap them as anonymous tag-capable mappings, re-loading content
+from the file where necessary.
+
+The ``DT_AARCH64_MEMTAG_GLOBALS`` dynamic table entry points to a
+``SHT_AARCH64_MEMTAG_GLOBALS_DYNAMIC`` section of compressed global variable
+descriptors. The size of this section, in bytes, is given by the
+``DT_AARCH64_MEMTAG_GLOBALSSZ`` dynamic table entry. When loading a shared
+object (including the main executable), prior to processing relocations, the
+dynamic loader should decompress the descriptors, and for each descriptor,
+perform the following operation:
+
+1. Generate a random tag (probably using the ``irg`` instruction). Ideally, to
+   ensure linear global buffer overflows are caught determinstally, the tag
+   should be different from adjacent global variables.
+
+2. Apply the random tag, using the ``stg`` family of instructions to the number
```

Comment:
```
2. Apply the random tag, using the ``stg`` family of instructions, to the number
   of granules specified in the descriptor.
```
Extra comma before "to the number". Or you could put "using the stg family of instructions" in brackets.
```
2. Apply the random tag (using the ``stg`` family of instructions) to the number
   of granules specified in the descriptor.
```